### PR TITLE
Replace const pointer with const reference in type functions

### PR DESCRIPTION
**Pull request by manh9203**
This PR resolved issue #3395 
The parameter type in ```ListType```, ```ArrayType```, ```StructType```, ```MapType```, and ```UnionType``` is modified to ```const LogicalType&```, and the return type is changed to ```LogicalType```.

### CODE REVIEW

File: src/binder/bind/read/bind_unwind.cpp [hunk 1]
```diff
@@ -21,12 +21,12 @@ std::unique_ptr<BoundReadingClause> Binder::bindUnwindClause(const ReadingClause
     auto aliasName = unwindClause.getAlias();
     std::shared_ptr<Expression> alias;
     if (boundExpression->getDataType().getLogicalTypeID() == LogicalTypeID::ARRAY) {
-        auto targetType = LogicalType::LIST(*ArrayType::getChildType(&boundExpression->dataType));
+        auto targetType = LogicalType::LIST(*ArrayType::getChildType(boundExpression->dataType));
```
Comment (acquamarin) [inline src/binder/bind/read/bind_unwind.cpp:24]: can we let ArrayType::getChildType() also return a reference rather a pointer

File: src/binder/expression_binder.cpp [hunk 2]
```diff
@@ -90,18 +90,18 @@ static bool compatible(const LogicalType& type, const LogicalType& target) {
     }
     switch (type.getLogicalTypeID()) {
     case LogicalTypeID::LIST: {
-        return compatible(*ListType::getChildType(&type), *ListType::getChildType(&target));
+        return compatible(*ListType::getChildType(type), *ListType::getChildType(target));
```
Comment (acquamarin) [inline src/binder/expression_binder.cpp:93]: Can we make getChildType() return reference as well?

File: src/c_api/value.cpp [hunk 3]
```diff
@@ -215,13 +215,13 @@ kuzu_value* kuzu_value_get_list_element(kuzu_value* value, uint64_t index) {
 uint64_t kuzu_value_get_struct_num_fields(kuzu_value* value) {
     auto val = static_cast<Value*>(value->_value);
     auto data_type = val->getDataType();
```
Comment (acquamarin) [inline src/c_api/value.cpp:217]: getDataType can also return a const reference?

File: src/common/types/types.cpp [hunk 4]
```diff
@@ -20,98 +20,98 @@ using kuzu::function::BuiltInFunctionsUtils;
...
+struct_field_idx_t StructType::getFieldIdx(const LogicalType& type, const std::string& key) {
+    KU_ASSERT(type.getPhysicalType() == PhysicalTypeID::STRUCT);
+    auto structTypeInfo = type.extraTypeInfo->constPtrCast<StructTypeInfo>();
     return structTypeInfo->getStructFieldIdx(key);
 }
 
-LogicalType* MapType::getKeyType(const LogicalType* type) {
-    KU_ASSERT(type->getLogicalTypeID() == LogicalTypeID::MAP);
-    return StructType::getFieldTypes(ListType::getChildType(type))[0];
+LogicalType* MapType::getKeyType(const LogicalType& type) {
```
Comment (acquamarin) [inline src/common/types/types.cpp:89]: Return reference?

File: src/common/types/types.cpp [hunk 5]
```diff
@@ -20,98 +20,98 @@ using kuzu::function::BuiltInFunctionsUtils;
...
-std::string UnionType::getFieldName(const LogicalType* type, union_field_idx_t idx) {
-    KU_ASSERT(type->getLogicalTypeID() == LogicalTypeID::UNION);
+std::string UnionType::getFieldName(const LogicalType& type, union_field_idx_t idx) {
+    KU_ASSERT(type.getLogicalTypeID() == LogicalTypeID::UNION);
     return StructType::getFieldNames(type)[getInternalFieldIdx(idx)];
 }
 
-LogicalType* UnionType::getFieldType(const LogicalType* type, union_field_idx_t idx) {
-    KU_ASSERT(type->getLogicalTypeID() == LogicalTypeID::UNION);
+LogicalType* UnionType::getFieldType(const LogicalType& type, union_field_idx_t idx) {
```
Comment (acquamarin) [inline src/common/types/types.cpp:108]: Ditto, it is better to return reference rather than pointer

**Review by acquamarin — APPROVED** [review]

**Review by andyfengHKU — APPROVED** [review]
LGTM

**Pull request merged**